### PR TITLE
[SPARK-10884][ML] Support prediction on single instance for regression and classification related models	

### PR DESCRIPTION
**Pull request by WeichenXu123**
## What changes were proposed in this pull request?

Support prediction on single instance for regression and classification related models (i.e., PredictionModel, ClassificationModel and their sub classes).
Add corresponding test cases.

## How was this patch tested?

Test cases added.

### CODE REVIEW

**Comment by drod331** [comment]
@WeichenXu123 My team and I are looking forward to the flexibility and efficiency provided by these changes.  What can I do to help get these changes merged into Spark master?  Thanks for your contribution.

**Comment by WeichenXu123** [comment]
cc @jkbradley @yanboliang thanks!

**Comment by smurching** [comment]
Looking at this now!

File: mllib/src/test/scala/org/apache/spark/ml/classification/DecisionTreeClassifierSuite.scala [hunk 1]
```diff
@@ -267,6 +268,24 @@ class DecisionTreeClassifierSuite
...
+      .setMaxBins(100)
+    val categoricalFeatures = Map(0 -> 3)
+    val numClasses = 3
+
+    val newData: DataFrame = TreeTests.setMetadata(rdd, categoricalFeatures, numClasses)
+    val newTree = dt.fit(newData)
+
+    newTree.transform(newData).select(dt.getFeaturesCol, dt.getPredictionCol).collect().foreach {
+      case Row(features: Vector, prediction: Double) =>
+        assert(prediction ~== newTree.predict(features) relTol 1E-5)
```
Comment (smurching) [inline mllib/src/test/scala/org/apache/spark/ml/classification/DecisionTreeClassifierSuite.scala]: Can we test exact equality (e.g. `prediction === newTree.predict(features)`) here and in other unit tests?

**Review by smurching — COMMENTED** [review]
Thanks for the changes! Had just one comment about the tests. Also, can you add a unit test in `NaiveBayesSuite` resembling those you've added for other models?

File: mllib/src/test/scala/org/apache/spark/ml/classification/NaiveBayesSuite.scala [hunk 2]
```diff
@@ -165,6 +165,35 @@ class NaiveBayesSuite extends SparkFunSuite with MLlibTestSparkContext with Defa
...
+    ).map(_.map(math.log))
+    val pi = Vectors.dense(piArray)
+    val theta = new DenseMatrix(3, 4, thetaArray.flatten, true)
+
+    val testDataset =
+      generateNaiveBayesInput(piArray, thetaArray, nPoints, seed, "multinomial").toDF()
+    val nb = new NaiveBayes().setSmoothing(1.0).setModelType("multinomial")
+    val model = nb.fit(testDataset)
+
+    validateModelFit(pi, theta, model)
```
Comment (smurching) [inline mllib/src/test/scala/org/apache/spark/ml/classification/NaiveBayesSuite.scala]: Do we need lines 184-186? They seem unrelated to what we want to test (that `predict` produces the same result as `transform` on a single instance). Similarly, I don't think we need to create `piArray`, `thetaArray`, `pi`, `theta`, etc; this test should just fit a model on a dataset and compare the fitted model's `predict` and `transform` outputs.

File: mllib/src/test/scala/org/apache/spark/ml/classification/NaiveBayesSuite.scala [hunk 3]
```diff
@@ -165,6 +165,35 @@ class NaiveBayesSuite extends SparkFunSuite with MLlibTestSparkContext with Defa
...
+    val piArray = Array(0.5, 0.1, 0.4).map(math.log)
+    val thetaArray = Array(
+      Array(0.70, 0.10, 0.10, 0.10), // label 0
+      Array(0.10, 0.70, 0.10, 0.10), // label 1
+      Array(0.10, 0.10, 0.70, 0.10)  // label 2
+    ).map(_.map(math.log))
+    val pi = Vectors.dense(piArray)
+    val theta = new DenseMatrix(3, 4, thetaArray.flatten, true)
+
+    val testDataset =
```
Comment (smurching) [inline mllib/src/test/scala/org/apache/spark/ml/classification/NaiveBayesSuite.scala]: Suggestion: rename to `trainDataset`?

File: mllib/src/test/scala/org/apache/spark/ml/classification/LinearSVCSuite.scala [hunk 4]
```diff
@@ -202,6 +202,15 @@ class LinearSVCSuite extends SparkFunSuite with MLlibTestSparkContext with Defau
       dataset.as[LabeledPoint], estimator, modelEquals, 42L)
   }
 
+  test("prediction on single instance") {
+    val trainer = new LinearSVC()
+    val model = trainer.fit(smallBinaryDataset)
+    model.transform(smallBinaryDataset).select("features", "prediction").collect().foreach {
+      case Row(features: Vector, prediction: Double) =>
+        assert(prediction ~== model.predict(features) relTol 1E-5)
```
Comment (smurching) [inline mllib/src/test/scala/org/apache/spark/ml/classification/LinearSVCSuite.scala]: Could you please check for exact equality here (& in other tests)?

**Review by smurching — COMMENTED** [review]
Thanks for the update, made a few comments!

File: mllib/src/test/scala/org/apache/spark/ml/regression/RandomForestRegressorSuite.scala [hunk 5]
```diff
@@ -19,14 +19,16 @@ package org.apache.spark.ml.regression
 
 import org.apache.spark.SparkFunSuite
 import org.apache.spark.ml.feature.LabeledPoint
+import org.apache.spark.ml.linalg.Vector
 import org.apache.spark.ml.tree.impl.TreeTests
 import org.apache.spark.ml.util.{DefaultReadWriteTest, MLTestingUtils}
+import org.apache.spark.ml.util.TestingUtils._
```
Comment (smurching) [inline mllib/src/test/scala/org/apache/spark/ml/regression/RandomForestRegressorSuite.scala]: Nit: unused import

File: mllib/src/test/scala/org/apache/spark/ml/regression/GBTRegressorSuite.scala [hunk 6]
```diff
@@ -19,15 +19,16 @@ package org.apache.spark.ml.regression
 
 import org.apache.spark.SparkFunSuite
 import org.apache.spark.ml.feature.LabeledPoint
-import org.apache.spark.ml.linalg.Vectors
+import org.apache.spark.ml.linalg.{Vector, Vectors}
 import org.apache.spark.ml.tree.impl.TreeTests
 import org.apache.spark.ml.util.{DefaultReadWriteTest, MLTestingUtils}
+import org.apache.spark.ml.util.TestingUtils._
```
Comment (smurching) [inline mllib/src/test/scala/org/apache/spark/ml/regression/GBTRegressorSuite.scala]: Nit: unused import

**Review by smurching — COMMENTED** [review]
Just a few nits; there are some unused imports now that we check for exact equality in tests. Otherwise, this LGTM! @yanboliang would you be able to take a look?

**Comment by jkbradley** [comment]
I'll try to take a look but am pretty swamped currently.  CC @yanboliang @MLnick @dbtsai @holdenk  might you have time?

**Comment by WeichenXu123** [comment]
Jenkins, test this please.

**Comment by WeichenXu123** [comment]
Jenkins, test this please.

**Comment by jkbradley** [comment]
Thanks for the PR!
Can you please add Since annotations?
Also, can the test code be consolidated?  Maybe you can have a helper function taking a model and dataset, usable with all Predictors.

File: mllib/src/test/scala/org/apache/spark/ml/util/MLTest.scala [hunk 7]
```diff
@@ -109,4 +110,14 @@ trait MLTest extends StreamTest with TempDirectory { self: Suite =>
     testTransformerOnDF(dataframe, transformer, firstResultCol,
       otherResultCols: _*)(globalCheckFunction)
   }
+
+  def testPredictorModelSinglePrediction(model: PredictionModel[Vector, _],
```
Comment (jkbradley) [inline mllib/src/test/scala/org/apache/spark/ml/util/MLTest.scala]: Thanks for the updates!  Just this comment: Can we please call this method testPredictionModelSinglePrediction to match the name of the class it is testing?

**Comment by jkbradley** [comment]
LGTM
Merging after fresh tests run
Thanks a lot @WeichenXu123 for the PR and others for reviews!

**Comment by WeichenXu123** [comment]
Jenkins, retest this please.

**Comment by jkbradley** [comment]
Merging with master

File: mllib/src/main/scala/org/apache/spark/ml/classification/Classifier.scala [hunk 8]
```diff
@@ -192,12 +192,12 @@ abstract class ClassificationModel[FeaturesType, M <: ClassificationModel[Featur
...
   /**
    * Predict label for the given features.
-   * This internal method is used to implement `transform()` and output [[predictionCol]].
+   * This method is used to implement `transform()` and output [[predictionCol]].
    *
    * This default implementation for classification predicts the index of the maximum value
    * from `predictRaw()`.
    */
-  override protected def predict(features: FeaturesType): Double = {
+  override def predict(features: FeaturesType): Double = {
```
Comment (zhengruifeng) [inline mllib/src/main/scala/org/apache/spark/ml/classification/Classifier.scala]: Is @Since annotation needed here? @WeichenXu123

Comment (WeichenXu123) [reply mllib/src/main/scala/org/apache/spark/ml/classification/Classifier.scala]: I think no. It will inherit `@Since` from super class `predict` method.

**Comment by dbtsai** [comment]
Too late to the party! Great PR. We should also think about refactor the predictors to `mllib-local` eventually so people can use them in prod without depending on Spark.

**Comment by WeichenXu123** [comment]
@dbtsai Good idea! Is there a related JIRA or could you open one for it ?
cc @jkbradley

**Comment by dbtsai** [comment]
@WeichenXu123 we were discussing this when we moved out common math code into `mllib-local`, but there is no umbrella ticket around it. I talked to many companies, and that is one of the pain point using mllib. Feel free to create ticket. 

Thanks.
